### DISCUSSION TITLE
Integrate with MLIR

### DISCUSSION THREAD

**Discussion by saulshanabrook** [Quansight-Labs/metadsl#106] [Ideas]
This project is effectively a Python-first solution for the same problem that MLIR faces, a a unified framework for using a number of competing and useful IRs for ML/DS.

In Python land, the goal is not to "pick the best IR/framework" for all users (i.e. pytorch, tf, etc) or to unify to a single shared IR that is the minimum common subset of all those (this is what the [Consortium for Python Data API Standards](https://data-apis.org/ is up to), but to provide users with a common way for not only *using* APIs/IRs but defining them and defining how they are compiled into one another.

The underlying goal being democratization of the compiler pipeline, to remove some of the barrier between being a library user, library author, and library optimizer. Politically, this is also a way to make sure that it's not only the companies with the largest resources who are able to create Python libraries to capture audiences for their hardware or clouds, but that users can have the freedom to be agnostic to their backend and switch when they like.

---

I recently skimmed a *great* paper on the subject, called ["The Deep Learning Compiler: A Comprehensive Survey"](https://arxiv.org/abs/2002.03794). In their conclusion they note:

> **Unified optimizations** - Although existing DL compilers adopt similar designs in both computation graph optimizations and hardware-specific optimizations, each compiler has its own advantages in certain aspects. There is a missing way to share the state-of-the-art optimizations, as well as support of emerging hardware targets across existing compilers. We advocate unifying the optimizations from existing DL compilers so that the best practices adopted in each DL compiler can be reused. In addition, unifying the optimizations across DL compilers can accumulate a strong force to impact the design of general-purpose and dedicated DL accelerators, and provide an environment for efficient co-design of DL compiler and hardware.
> 
> Currently, Google MLIR is a promising initiative towards such direction. It provides the infrastructure of multi-level IRs, and contains IR specification and toolkit to perform transformations across IRs at each level. It also provides flexible dialects, so that each DL compiler can construct its customized dialects for both high-level and low-level IRs. Through transformation across dialects, optimizations of one DL compiler can be reused by another compiler. However, the transformation of dialects requires further research efforts to reduce the dependency on delicate design.

So if we can create some interop with MLIR, treating the UX of metadsl as a goal, combined with the existing ecosystem of MLIR, we can empower users from Python to use and create data science compilers.

---

The [npcomp](https://github.com/llvm/mlir-npcomp) project is already trying to make some Python integration for MLIR. When I last chatted with @stellaraccident many months ago, I talked about some parts that I would be interested in seeing added on top of the work there, including:

* Ability to take an existing MLIR dialect and auto generate a deeply embedded Python DSL for it, with complete typings. To do this, we need to make sure we need to investigate how much of MLIR's typing semantics can be translated to Python's (in the static/mypy world). From first glance before.
* Ability to write Python code (classes, functions, etc) to define a new dialect.
* Ability to write Python code to define ways to convert between dialects
* Add standard way to "execute" some compiled code from MLIR inside of Python.

Basically, the goal here is now to necessarily even build a new dialect for anything Python related, at least at first, but to allow other library authors to use the DSLs of existing dialects from Python to build their own compiler toolchains.

**Comment by stellaraccident** [comment]
Hi - nice to chat again!

There have been some developments since we last spoke. Namely, since then, we've made progress on upstream Python bindings that are starting to edge towards some of the points you mention. We've been trying to make them a good starting point for constructing nice DSLs from (versus trying to be a good DSL for everything):

  * Exposes the core IR constructs (context, modules, ops, types, attributes).
  * Supports with blocks and context managers for implicit management of contexts, locations and insertion points.
  * Integrated with the ODS infra so that we can generate Python wrappers for each dialect. 
  * Run passes.

Code: https://github.com/llvm/llvm-project/tree/master/mlir/lib/Bindings/Python
Docs: https://mlir.llvm.org/docs/Bindings/Python/ (slightly out of date)
Tests: https://github.com/llvm/llvm-project/tree/master/mlir/test/Bindings/Python (more up to date than docs. [Example](https://github.com/llvm/llvm-project/blob/master/mlir/test/Bindings/Python/dialects/std.py) of exercising the std dialect via (very new) autogenerated Python wrapper)

There are still things that are in progress:

* Custom attributes and types have been discussed but integration with Python still requires native code
* Need to wire in the execution engine/JIT interface (npcomp has a special purpose thing for this - need to implement the proper thing upstream)
* Deployability/packaging and integration with other projects needs more work. [Meeting notes](https://llvm.discourse.group/t/meeting-notes-mlir-build-install-and-shared-libraries/2257) where some of the necessary work was planned.

I am also aware of at least three other developments happening towards some of your other points but in various state of work/disclosure. I think @joker-eph (Mehdi) is a good join point with visibility of all of those. Boosting this to a thread on the [LLVM/MLIR Discourse](https://llvm.discourse.group/c/mlir/31) may be a good way to have the discussion in view of some of those things.
